csrc/tensor_metadata.cpp

```diff
@@ -259,6 +260,24 @@ void validateAllocationSizesAndStrides(
           [](auto c_flag) { return c_flag.has_value(); }),
       "The size of contiguity mismatch with the dimensionality of allocation domain");
 
+  // Validate that strides must be in decreasing order modulo reduction and
```

Comment:
Why strides must be decreasing? If we have a discontiguous tensor, can't it be arbitrary?

Reply:
IIUC, the input `sizes` and `strides` are rearranged according to the allocation domain and therefore in major-to-minor order. So I'm expecting `strides` to be in decreasing order modulo broadcast and reduction. Otherwise, it's not compatible with the allocation domain.

Reply:
I think:
1. Fusion executor cache should reorder it as you described.
2. Even if fusion executor does not reorder it, as long as it is marked as discontinuity, the rest of our system should support it.

So I do think we need such a test, but not here.

Reply:
I think we might have talked past each other. The reorder happens [here](https://github.com/NVIDIA/Fuser/blob/b4ef869c36de5b36c67f3f118c59012d618b434f/csrc/tensor_metadata.cpp#L318) in `inferAndValidateAllocationSizesAndStrides`. `validateAllocationSizesAndStrides` is an internal function whose input `sizes` and `strides` are guaranteed to be reordered (and thus my expectation). Does that make sense?

Reply:
Are we talking about the same reorder? I think there are two different "reorder"s we can think about. One reorder is to reorder the computed sizes and strides to be consistent with the allocation domain. This reorder does not change allocation domain itself. For this reordering, it happens in the code you refers to. But the "reorder" I was referring to is to set the allocation domain so that the order of the allocation domain matches with the stride order of the ATen tensor. This is a process similar to how we set the contiguity of a tensor.

Reply:
Got it. This goes back to my question 3 in OP. AFAICS, @jjsjann123 proposed that an empty allocation domain means no constraint, which I agree. So here I shouldn't expect sizes and strides to be in certain order when the allocation domain is empty. Sgty?

Reply:
> I shouldn't expect sizes and strides to be in certain order in certain order when the allocation domain is empty.

I think we shouldn't expect sizes and strides to be in certain order in all cases. For example, if my allocation domain is `[I1, I2]` and my stride is `[1, 10]`, although this is not performant, I don't see any reason why we should reject this. This can be handled as a general discontiguous tensor. It's not the most performant kernel, but it is functionally correct. I think our lowering and execution should be tolerant to all good and bad schedules, having a non-decreasing stride order is a bad schedule, but not wrong. This is just like C++ should support all good and bad coding styles, passing by value instead of by `const &` can be slow, but the C++ compiler should not reject the program.

Reply:
I consider nvFuser as two parts:
1. Fusion, lowering, executor - a programming language like C++
2. Schedulers, executor cache - an automatic code generator that generates programs written in the above language

Reply:
> For example, if my allocation domain is [I1, I2] and my stride is [1, 10], although this is not performant, I don't see any reason why we should reject this. 

I'm confused. What's the purpose of allocation domain if it doesn't dictate a physical memory layout (and therefore a stride order)?

Reply:
Xiang seems to be suggesting us to relax what `allocDomain` means. i.e. `allocDomain` is only semantically meaningful when we have the corresponding contiguity flag set as `true`?

I felt this makes some sense from the design of indexing. nitpicking on this topic,  it felt a bit sloppy conceptually. Given that `allocDomain` and `contiguity` are two separate field. But our API only allows the mutation of both field together, so maybe we are not that far away from the reality that `allocDomain` and `contiguity` should come as a single package :shrug: 

both protocol sound fine to me.

Reply:
> allocDomain is only semantically meaningful when we have the corresponding contiguity flag set as true?

Interesting. Does that mean `allocation_domain[i]` is completely useless when `contiguity[i]==false`?

Reply:
> What's the purpose of allocation domain if it doesn't dictate a physical memory layout (and therefore a stride order)?

It does dictate the physical memory layout, but "dictate the physical memory layout" can not be simply translated to "require the stride order to be decreasing". For example, if I have a tensor whose rFactor domain is `[I1, I2, I3]`, shape is `(4, 5, 6)`, and stride is `(6, 2, 1)`, then it might makes more sense to have allocation domain `[I2, I1, I3]` and contiguity `(false, true, true)`, instead of having allocation domain `[I1, I2, I3]` and contiguity `(false, false, true)`.

Reply:
> maybe we are not that far away from the reality that `allocDomain` and `contiguity` should come as a single package 🤷

I think they are already a single package, no?

Reply:
Re @zasdfgbnm 

> I think they are already a single package, no?

Conceptually I think yes. But looking at the implementation we have, it's likely a no... hence the confusion that Jingyue is having regarding the topic.

Re @wujingyue 

> >  allocDomain is only semantically meaningful when we have the corresponding contiguity flag set as true?
> 
> Interesting. Does that mean allocation_domain[i] is completely useless when contiguity[i]==false?

Sorta yes. You would also need `contiguity[i-1] == false`.
But this isn't a definitive answer there. As @zasdfgbnm points out, we can still rely on the order for scheduler, but that's should be just a performance hint.
I think Xiang's point here is that, we shouldn't assert on stride order when corresponding contiguity flags were set to false.
I agree with this ^^^. Even though in general, I'm not a big fan of  "be conservative in what you do, be liberal in what you accept from others" :laughing: